docs/architecture.md

```diff
@@ -0,0 +1,331 @@
+# Fuel Client Software Architecture
+
+Inspired by [Hexagonal Architecture](https://en.wikipedia.org/wiki/Hexagonal_architecture_(software)), the Fuel 
+blockchain client is composed of various domain modules. Each module defines its own view of the system 
+via a class of traits we'll call "Ports". This is in contrast to creating globally shared interfaces
+with the difficulty of trying to cover every possible use-case in a future-proof way, while on the other hand
+leaking many irrelevant concerns across different domains. This approach may lead to a small amount of
+monotonous and boring bloat in a few areas, but this is offset by the advantages of loose coupling.
+This is intended to minimize the overall complexity and cognitive load of the codebase.
+Business logic will have clear intentions and be easy to follow, unit tests don't need to mock low level details
+about unrelated components, and refactors can easily control their blast radius.
+
+In a system of shared global traits between modules, any updates to an interface will cause breaking changes
+to the mocks and test harnesses for all consuming modules. This is less than ideal when these domain modules
+don't need to use the new changes. However, in a system where each domain controls its ports, adjusting an
+interface will only affect the port owner and the implementor.
+
+The implementation of "Ports" (aka adapters) referring to other domain modules, should reside in the main fuel-core
+crate. This way, domain modules should never need to take a dependency on each other. This helps to avoid circular
+dependencies, and also allows for lightweight rapid developer iteration within a domain module. It also ensures
+the purity of the domain modules, preventing implementation details of modules from contaminating each other. Adapters
+are intended to restrict the viral spread of any ugly complexity that arises when integrating modules. Decisions
+such as whether to use channels, networking, IPC, or direct calls between domain modules can be decided in the adapter
+without any effect on the domain logic. They should avoid implementing any business rules, and only contain integration
+glue logic between systems.
+
+Special care was also taken in this design approach to keep implementation details regarding synchronization primitives
+or proprietary async runtime requirements out of any public interfaces (No channels!).
+
```

Comment:
It will be good to add a description that all primary implementation of those traits should be done in the corresponding services and reused in the `fuel-core`. Because our main goal is that everyone can take any of our services(separate crate), and it should provide most of the required functionality.

Example:

If we need the list of all consensus participants from the database, it should already have a public API that returns it.

Or

If we need to implement `broadcast_transaction`, P2P should provide some method `broadcast`, and we can reuse it in the `fuel-core`.

Because implementation of all business logic in the `fuel-core` will make our crate system useless then=)

Reply:
added the following guidelines about adapters:

- Adapters implement Ports. They are for integrating with the "real-world" dependencies like using other domain modules,
   implementing serialization logic, or calling a database.
 - Adapters should either reside in the top-level crate ("fuel-core") or in their own crates. This prevents the domains
   from being polluted by dependencies with large compile times or unstable interfaces. This allows for libraries like 
   networking, third-party SDKs, or database drivers to be swapped out without breaking our core domain specific logic.